app/models/ssh_execution_provider.rb

```diff
@@ -18,27 +18,58 @@ def supports_effective_user?
       true
     end
 
+    def find_ip_or_hostname(host)
+      interfaces = effective_interfaces(host)
+      if host_setting(host, :remote_execution_connect_by_ip)
+        ip_interface = interfaces.find { |i| i.ip.present? }
+      end
+      if ip_interface
+        ip_interface.ip
+      else
+        fqdn_interface = interfaces.find { |i| i.fqdn.present? }
+        if fqdn_interface
+          fqdn_interface.fqdn
+        else
+          host.fqnd
```

Comment:
(Although this `else` could probably go away, a host must always have a primary interface, and host.fqdn is delegated to it.)